mitmproxy/addons/maplocal.py

```diff
@@ -1,52 +1,86 @@
 import mimetypes
 import re
 import typing
-import urllib
 from pathlib import Path
+
 from werkzeug.security import safe_join
 
-from mitmproxy import ctx, exceptions, http
-from mitmproxy.addons.modifyheaders import parse_modify_spec, ModifySpec
+from mitmproxy import ctx, exceptions, flowfilter, http
+from mitmproxy.addons.modifyheaders import parse_spec
+
+
+class MapLocalSpec(typing.NamedTuple):
+    matches: flowfilter.TFilter
+    regex: str
+    local_path: Path
+
+
+def parse_map_local_spec(option: str) -> MapLocalSpec:
+    filter, regex, replacement = parse_spec(option)
+
+    try:
+        re.compile(regex)
+    except re.error as e:
+        raise ValueError(f"Invalid regular expression {regex!r} ({e})")
+
+    try:
+        path = Path(replacement).expanduser().resolve(strict=True)
+    except FileNotFoundError as e:
+        raise ValueError(f"Invalid file path: {replacement} ({e})")
+
+    return MapLocalSpec(filter, regex, path)
 
 
 def get_mime_type(file_path: str) -> str:
     mimetype = (
-        mimetypes.guess_type(file_path)[0]
-        or "application/octet-stream"
+            mimetypes.guess_type(file_path)[0]
+            or "application/octet-stream"
     )
     return mimetype
 
 
-def file_candidates(url: str, base_path: str) -> typing.List[Path]:
-    candidates = []
-    parsed_url = urllib.parse.urlparse(url)
-    path_components = parsed_url.path.lstrip("/").split("/")
-    filename = path_components.pop()
+def _safe_path_join(root: Path, untrusted: str) -> Path:
+    """Join a Path element with an untrusted str.
 
-    # todo: we may want to consider other filenames such as index.htm)
-    if not filename:
-        filename = 'index.html'
+    This is just a convenience wrapper for werkzeug's safe_join."""
+    untrusted_parts = Path(untrusted).parts
+    joined = safe_join(
+        root.as_posix(),
+        *untrusted_parts
+    )
+    if joined is None:
+        raise ValueError("Untrusted paths.")
+    return Path(joined)
 
-    # construct all possible paths
-    while True:
-        components_with_filename = tuple(path_components + [filename])
-        candidate_path = safe_join(base_path, *components_with_filename)
-        if candidate_path:
-            candidates.append(
-                Path(candidate_path)
-            )
 
-        if not path_components:
-            break
+def file_candidates(url: str, spec: MapLocalSpec) -> typing.List[Path]:
+    candidates = []
 
-        path_components.pop()
+    m = re.search(spec.regex, url)
+    assert m
+    if m.groups():
+        suffix = m.group(1)
+    else:
+        suffix = re.split(spec.regex, url, maxsplit=1)[1]
```

Comment:
Here you assume that the user provides a regex that starts with the path (and an optional hostname).
Assumeing the URL is `https://example/foo/bar/baz.jpg` and subject is `:/bar/baz:/tmp`, then the determined suffix is`.jpg`, which does not make sense. Are we ok with that and state in the docs that the regex should start with `hostname/path-prefix`?

Reply:
I think the best way to document that is state that the pattern splits the URL, and the part right of the pattern is appended to the directory. It would be cool to have two simple colored examples like this:

```
HTTP Request:  https://example.com/foo/bar/qux.jpg?timestamp=123
                              ▲             │
URL Regex: example.com/foo   ─┘             │
                                            ▼
Served File:                  ~/static/bar/qux.jpg
```

```
HTTP Request:  https://example.com/view?id=1
                                 ▲     │
URL Regex: example.com/(.+)     ─┘     │
                                       ▼
Served File:              ~/static/view_id=1
```

Unfortunately I can't color here, but hopefully it gets the point across. :-)

Reply:
Talking about examples:

- we should probably list one example with `:~m GET:` to illustrate the point of the flow filter.
- We could consider discussing regex lookaheads here. I'm not sure what's simpler, do we prefer `:example.com/(?=.+\.jpg$):~/images` or `:~u .jpg$:example.com:~/images`? I guess both would get the job done. :)

Reply:
Good idea. For your second example, the regex should be `example.com/` or `example.org` I think. Will put a colored ASCII example in the docs, or did you think of an image?
Regarding examples: I think lookaheads often make regexes hard to read. Using the flow filter might be confusing as well, the user might wonder what's the purpose of the url-regex. But yes, both variants get the job done. Will think about it.